### PR TITLE
ValidateJoinDepth rule; hard cutoff for > 12 table joins

### PR DESCRIPTION
**Pull request by max-hoffman**
New analyzer rule `validateJoinDepth` that errors if we count more than 12 join leaves. Opaque node counting could be done more carefully.

join search schema comparison was O(n^4) if i counted correctly, now should be more like O(n)

### CODE REVIEW

File: sql/analyzer/indexed_joins.go [hunk 1]
```diff
@@ -112,6 +120,55 @@ func replaceJoinPlans(ctx *sql.Context, a *Analyzer, n sql.Node, scope *Scope) (
 	return withIndexedTableAccess, transform.NewTree, nil
 }
 
+// countJoinDepth uses a naive algorithm to count
+// the number of join leaves in a query.
+//todo(max): recursive ctes with joins might be double counted,
+// tricky to test
+func countJoinDepth(n sql.Node) int {
```
Comment (zachmu) [inline sql/analyzer/indexed_joins.go:127]: depth is deceptive here

`countTableFactors`

File: sql/analyzer/indexed_joins.go [hunk 2]
```diff
@@ -536,14 +597,9 @@ func (j joinIndexes) getUsableIndex(schema sql.Schema) *joinIndex {
...
-func schemaContainsField(schema sql.Schema, field *expression.GetField) bool {
-	for _, col := range schema {
-		if strings.ToLower(col.Source) == strings.ToLower(field.Table()) &&
-			strings.ToLower(col.Name) == strings.ToLower(field.Name()) {
-			return true
-		}
-	}
-	return false
+func schemaContainsField(schemaCols map[tableCol]struct{}, field *expression.GetField) bool {
+	_, ok := schemaCols[tableCol{field.Table(), field.Name()}]
```
Comment (zachmu) [inline sql/analyzer/indexed_joins.go:601]: You are omitting the strings.ToLower here, are you sure it's applied higher in the stack?

File: sql/analyzer/join_search.go [hunk 3]
```diff
@@ -24,7 +24,7 @@ import (
 	"github.com/dolthub/go-mysql-server/sql/plan"
 )
 
-const searchLimit = 1 << 29 // includes 12 tables joins
+const joinCountLimit = 12
```
Comment (zachmu) [inline sql/analyzer/join_search.go:27]: `joinTablesLimit`

File: enginetest/queries.go [hunk 4]
```diff
@@ -8569,6 +8569,81 @@ var errorQueries = []QueryErrorTest{
...
+						mytable as t10,
+						mytable as t11,
+						mytable as t12,
+						mytable as t13,
+						mytable as t14,
+						mytable as t15`,
+		ExpectedErr: sql.ErrUnsupportedJoinCount,
+	},
+	{
+		Query: `SELECT t1.*
```
Comment (zachmu) [inline enginetest/queries.go:8628]: Make sure we don't error out in analysis when joining e.g. 2 subqueries that have 10 tables each

**Review by zachmu — APPROVED** [review]
Generally fine, see comments

File: enginetest/queries.go (hunk 4 above)
Comment (max-hoffman) [reply enginetest/queries.go:8628]: yeah that's a good point, didn't think of that. i guess opaque nodes should be included in the outer scope table count, but have their children counted independently.

Comment (max-hoffman) [reply enginetest/queries.go:8628]: on second though, inserting subqueries in a join doesn't affect the join complexity, right? ex: a 3 table join with subqueries inserted is still a 3 table join. same is true for 20 table join.
```SQL
> select * from (select * from a) as a, (select * from b) as b , (select * from c) as c;
```

**Pull request merged**